### DISCUSSION TITLE
How to debug Vite app (Sveltekit) running on Bun?

### DISCUSSION THREAD

**Discussion by pjeziorowski** [oven-sh/bun#14759] [Q&A]
Hi!

Following the docs, I'm able to debug my backend app in VSCode (https://bun.sh/guides/runtime/vscode-debugger) using the `Debug File` option.

However, my other app is using Sveltekit + Vite with a `vite dev` to run the app for development, which makes it a bit trickier to attach the debugger the same way as for the backend app.

Has anyone managed to debug Vite apps running on bun?

**Comment by TomasRJ** [comment]
I found a half solution to this question, meaning i can't get breakpoints to work for server side typescript code, but they do work for typescript code blocks in `.svelte` files. I don't know what the problem is here [since it seems like you can debug server side code with npx](https://stackoverflow.com/a/78597252), probably since Svelte uses Vite, which has not ["been adapted to use Bun's bundler, module resolver, or transpiler"](https://bun.sh/guides/ecosystem/vite).

1. Edit your `package.json` files launch profiles like is shown here: https://bun.sh/guides/ecosystem/vite
    - If `bunx` is not found, then use `bun x`. Example:  
        ```json
            "dev": "bun x --bun vite",
            "build": "bun x --bun vite build",
            "preview": "bun x --bun vite preview",
        ```
2. Then you can use that new launch profile for the VSCode `launch.json` file like so:
    ```json
    "configurations": [
        {
            "command": "bun run dev",
            "name": "Bun attach",
            "request": "launch",
            "type": "node-terminal"
        }
    ]
    ```
3. Then you should be able to launch your Bun + Sveltekit project in debug mode.

**Reply by quansie** [reply]
I've been using this setup to debug SvelteKit server code running on Bun, and it's been working pretty well. Here's a step-by-step guide:


1. Create a `debug.ts` file in your project root with the code below
2. Run `bun --inspect debug.ts` to start the Vite dev server in debug mode
3. Bun will print a Web debugger URL in the console
4. Open that URL in Chrome (important: Chrome works, Safari doesn't due to security restrictions)
5. Now you can set breakpoints in your server code and debug away!

```typescript
import { createServer, type ViteDevServer, type InlineConfig } from "vite";

async function startViteDebug(): Promise<void> {
  console.log("Initializing Vite server in debug mode...");

  try {
    // Define Vite configuration with TypeScript types
    const config: InlineConfig = {
      // Point to SvelteKit's Vite config file
      configFile: "./vite.config.ts", // SvelteKit uses vite.config.ts or vite.config.js

      // Server configuration
      server: {
        port: 5173, // Change this if needed
        strictPort: true, // Fail if port is already in use
        hmr: false,
      },

      // Log level can be adjusted for debugging needs
      logLevel: "info",
    };

    // Create the Vite server with your configuration
    const server: ViteDevServer = await createServer(config);

    // Start the server
    await server.listen();

    // Output success message with URLs
    const info = server.config.server;
    console.log(`\nVite server started in debug mode!`);
    console.log(`➜ Local:   http://localhost:${info.port}/`);
  } catch (error) {
    console.error("Failed to start Vite server in debug mode:", error);
  }
}

// Actually call the function to start the server!
startViteDebug();
```

Note that HMR is turned off in this configuration, so you'll need to restart the server after any code changes. I had some issues with HMR on, but you may try to see if it works.

Hope this helps!

**Reply by FlorianB-DE** [reply]
Sadly this doesn't seem to work anymore:
```bash
Initializing Vite server in debug mode...
failed to load config from /app/vite.config.ts
Failed to start Vite server in debug mode: 1 | import _debug from 'debug';
2 | 
3 | export const debug = _debug('vite-plugin-svelte-inspector');
                         ^
TypeError: _debug is not a function. (In '_debug("vite-plugin-svelte-inspector")', '_debug' is an instance of Promise)
      at /app/node_modules/@sveltejs/vite-plugin-svelte-inspector/src/debug.js:3:22
```
